### PR TITLE
Extract resources with ZipFile from APK file directly

### PR DESCRIPTION
**Pull request by mimers**
AssetManager behave strangely on some devices(e.g. OD103 Smartisan): when reading content from `AssetManager.open("icudtl.dat")`, it will feed by the content from "system/usr/icu/icudt56l.dat".
This result can only be reproduced with my App which package name is `com.sankuai.meituan.takeoutnew`, flutter gallery and other demos I created will not be affected. I assume the OEM rom has different `AssetManager`s for some famous App. So extract resources with via Zip file can be more reliable.
This fixes https://github.com/flutter/flutter/issues/16993

### CODE REVIEW

**Comment by mimers** [comment]
I signed it!

File: shell/platform/android/io/flutter/view/ResourceExtractor.java [hunk 1]
```diff
@@ -38,40 +39,47 @@ private void extractResources() {
...
+                            continue;
+                        }
+                        if (output.getParentFile() != null) {
+                            output.getParentFile().mkdirs();
+                        }
 
-                            int count = 0;
-                            while ((count = is.read(buffer, 0, BUFFER_SIZE)) != -1) {
-                                os.write(buffer, 0, count);
+                        ZipEntry entry = apk.getEntry("assets/" + asset);
```
Comment (chinmaygarde) [inline shell/platform/android/io/flutter/view/ResourceExtractor.java]: I believe this directory name is configurable and not always guaranteed to the called "assets".

**Comment by jason-simmons** [comment]
Yes - I'd also like to understand why the Android AssetManager is behaving differently on the affected device.

Apps are generally expected to obtain assets through the AssetManager API and not make assumptions about the layout of the APK archive.

The documentation for [Context.getPackageCodePath](https://developer.android.com/reference/android/content/Context.html#getPackageCodePath()) notes that "this is not generally useful for applications, since they should not be directly accessing the file system"

File: shell/platform/android/io/flutter/view/ResourceExtractor.java (hunk 1 above)
Comment (mimers) [reply shell/platform/android/io/flutter/view/ResourceExtractor.java]: https://android.googlesource.com/platform/frameworks/base/+/master/libs/androidfw/AssetManager.cpp#438
The `kAssetsRoot` is constant value defined https://android.googlesource.com/platform/frameworks/base/+/master/libs/androidfw/AssetManager.cpp#62 . Android doesn't provide any chance to change the value.

**Comment by mimers** [comment]
@jason-simmons  I totally understand what you are concern about, and I can't believe this at the first time. 
[Smartisan](https://t.tt) is an Android phone brand which proud of their 'deeply customized' UI, UE and ID, and I believe the redirecting stream from AssetManager is a BUG, because it breaks the default behavior of Android. I have reported an issue to them, but unofficially. Anyway,  I must solve the crash before we release the app with flutter embedded. So I followed the suggest you replied to me at https://groups.google.com/d/msg/flutter-dev/ue2KFqf8dLo/vcChaNe7BQAJ .

The doc of getPackageCodePath said 'this is not `generally` useful for applications', but I think this is not a generally situation.

On the other way, we can avoid the contaminated AssetManager by changing all Flutter resources location. For example we can put a common prefix for all resources like 'flutter_' or put them in an directory like 'asset/flutter/*', exposed under the root asset directory is not `safe` on certain devices, especially with a common name(icudtl.dat).

**Comment by jason-simmons** [comment]
On the master branch of Flutter, APKs should have assets placed in `assets/flutter_assets` and `assets/flutter_shared` paths

**Comment by mimers** [comment]
I have noticed that after upgrade to 0.6 beta, this can be solved as well.
duplicated with #18514.

**Pull request closed without merging**